mist/worker/src/main/scala/io/hydrosphere/mist/worker/WorkerActor.scala

```diff
@@ -1,149 +1,145 @@
-      }
-      self ! message
-    })
-    jobStart
-  }
-
-  private def downloadFile(actor: ActorRef, req: RunJobRequest): Future[SparkArtifact] = {
-    actor ! JobFileDownloading(req.id)
-    artifactDownloader.downloadArtifact(req.params.filePath)
-  }
-
-  private def failure(req: RunJobRequest, ex: Throwable): JobResponse =
-    JobFailure(req.id, buildErrorMessage(req.params, ex))
-
-
-  private def runJob(actor: ActorRef, req: RunJobRequest, runner: JobRunner): Either[Throwable, JsLikeData] = {
-    val id = req.id
-    actor ! JobStarted(id)
-    namedContext.sparkContext.setJobGroup(id, id)
-    runner.run(req, namedContext)
-  }
-
-  protected def buildErrorMessage(params: JobParams, e: Throwable): String = {
-    val msg = Option(e.getMessage).getOrElse("")
-    val trace = e.getStackTrace.map(e => e.toString).mkString("; ")
-    s"Error running job with $params. Type: ${e.getClass.getCanonicalName}, message: $msg, trace $trace"
-  }
-}
 
 class WorkerActor(
   val runnerSelector: RunnerSelector,
   val namedContext: NamedContext,
   val artifactDownloader: ArtifactDownloader,
   mkAppender: String => Option[Appender]
-) extends Actor with JobStarting with ActorLogging {
-
-  implicit val ec = {
-    val logger = LoggerFactory.getLogger(this.getClass)
-    val service = Executors.newFixedThreadPool(1)
-    ExecutionContext.fromExecutorService(
-      service,
-      e => logger.error("Error from thread pool", e)
-    )
-  }
+) extends Actor with ActorLogging {
+
+  import context._
+
+  val sparkRootLogger = LogManager.getRootLogger
+
+  type JobFuture = CancellableFuture[JsLikeData]
 
   override def receive: Receive = awaitRequest()
 
   private def awaitRequest(): Receive = {
     case req: RunJobRequest =>
-      mkAppender(req.id).foreach(rootLogger.addAppender)
-      val jobStarted = startJob(req)
-      val unit = ExecutionUnit(sender(), jobStarted)
-      context become running(unit)
+      mkAppender(req.id).foreach(sparkRootLogger.addAppender)
+      sender() ! JobFileDownloading(req.id)
+      artifactDownloader.downloadArtifact(req.params.filePath) pipeTo self
+      context become downloading(sender(), req, running)
 
     case _: ShutdownCommand =>
       self ! PoisonPill
   }
 
-  private def running(execution: ExecutionUnit): Receive = {
+  private def downloading(
+    respond: ActorRef,
+    req: RunJobRequest,
+    next: (RunJobRequest, ActorRef, JobFuture) => Receive
+  ): Receive = {
+    case artifact: SparkArtifact =>
+      val runner = runnerSelector.selectRunner(artifact)
+      namedContext.sparkContext.setJobGroup(req.id, req.id)
+      val jobFuture = run(runner, req)
+      respond ! JobStarted(req.id)
+      jobFuture.future pipeTo self
+      context become next(req, respond, jobFuture)
+
+    case CancelJobRequest(id) =>
+      sparkRootLogger.removeAppender(id)
+      respond ! JobIsCancelled(id)
+      respond ! mkFailure(req, new RuntimeException("Job was cancelled before starting"))
```

Comment:
This message will not be received by `JobActor` because when `JobActor` receives `JobIsCancelled` it will send himself `PoisonPill`.

Reply:
That behavior was changed - `JobActor` anyway waits for success or failure message. Receiving `JobIsCancelled` means that jobs statuses can be updated onto `Cancelled`

Reply:
Oh, I was looking at the old version of the code. 
It looks like it works